Improve backend routes auth

alesshh opened the pull request:
# Problem
Our backend routes are unsafe and can be accessible without login. 😱 

## Solution
This PR will require user authentication on all private routes.

## Expected behavior

- The route `/` redirect users to rooms screen if already logged in
- The routes `/morpheus*` redirect unlogged users to login screen
- The route `/new` redirect unlogged users to login screen
- The route `/remove` redirect unlogged users to login screen
- The route `/rooms` send a 401 if the user is not logged in

**The socket layer remains unprotected because of we pass the user data to the socket, without verifications.**
The middleware used to store cookie sessions don't work with socket.io middlewares. 
We need to create a cookie session reader/writer, able to works with socket.io and express middlewares. I will open an issue to solve it in the future.

dvinciguerra commented on backend/app/app.routes.js:
```diff
@@ -1,25 +1,27 @@
+  }
+
+  return res.render("index", { error: req.query.error });
 });
```
Isn't it better to add a `home.controller.js` to handle this logic?

Something like this:

```javascript
import Home from './controllers/home.controller';

router.get("/", Home.index);
```

alesshh replied:
I think we can extract all route functions  on a separated pr. What do you think?

alesshh replied:
The index is the login page 😅 .
I don't like javascript classes for imports as much. My personal preference is to export only functions and explicit import what do you want.

What do you think to discuss about the backend structure / code on a issue?

dvinciguerra replied:
Perfect, I agree!!! 👏

dvinciguerra commented on backend/app/app.auth.js:
```diff
@@ -9,7 +9,7 @@ passport.use(
       clientSecret: config.GOOGLE_SECRET,
-      callbackURL: config.GOOGLE_CALLBACK_URL,
+      callbackURL: config.GOOGLE_CALLBACK_URL
     },
```
What do you think about this approach?

```javascript
import {
  GOOGLE_CLIENT_ID as clientID,
  GOOGLE_SECRET as clientSecret,
  GOOGLE_CALLBACK_URL as callbackURL
} from "./app.config";

...

passport.use(
  new GooglePassport.Strategy(
    { clientID, clientSecret, callbackURL },
    (accessToken, refreshToken, profile, cb) => {
		...
  })
}
```

alesshh replied:
Better! 🚀

dvinciguerra commented on backend/app/app.auth.js:
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
 });
 
```
What do you think about transform this object into a "module class" using static methods?

```javascript
class Auth {
  static initialize() { ... }

  static session() { ... }

  static isUserLoggedIn(req) { ... }

  static currentUser(req) { ... }

  static authenticate(opts) { ... }

  static login(req, user) { ... }

  static logout(req) { ... }
}

module.exports = Auth;
```

Look more organized in my opinion and don't change the usage (tests and imports).

alesshh replied:
I don't like javascript classes for imports as much. My personal preference is to export only functions and explicit import what do you want.

What do you think to discuss about the backend structure / code on a issue?

dvinciguerra replied:
Sure man! =)

megatroom replied:
I also prefer not to use classes in JavaScript, but either way doesn't make sense to me in having a class with only static methods, it run away from OOP.

megatroom commented on backend/app/app.auth.js:
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
+    return req.user;
+  },
+
+  authenticate(opts) {
```
What do you think about moving this method to a file inside a directory of middlewares? Example:

`backend/app/middlewares/authenticate.js`

Is not clear to me that this method is a middleware. I only found out when I saw the usage on the route.

What do you think?

megatroom commented on backend/app/app.auth.js:
```diff
@@ -32,8 +32,53 @@ passport.deserializeUser((user, done) => {
+  },
+
+  authenticate(opts) {
+    const { loginURL = null } = opts || {};
```
```js
const loginURL = opts && opts.loginURL;
```

What do you think of doing this way?

megatroom commented on frontend/src/pages/login/login.js:
```diff
@@ -14,85 +14,44 @@ import "bootstrap/dist/css/bootstrap.css";
-                  )}
-                </div>
-                <Footer />
+export function Login() {
```
:heart:

megatroom commented on frontend/src/pages/login/login.js:
```diff
@@ -14,85 +14,44 @@ import "bootstrap/dist/css/bootstrap.css";
-                </div>
-                <Footer />
+export function Login() {
+  const [error, setError] = useState(null);
```
:heart:

alesshh commented:
Hey guys, given your feedbacks I made some changes:
-  Extracted auth and passport responsibility to a service `app/services/auth` that encapsulate login and session passport logic, this way we can easily add new auth strategies with a single config.
- Refactored config file to functions,  this way, we do not need to use ugly constants on app setup, we can group configs by context;
- Added morgan logger to better development;
- Created an object to map route paths on route files;

angeliski commented on backend/app/app.server.js:
```diff
@@ -2,66 +2,61 @@ import express from "express";
 const app = express();
 
-app.locals.CUSTOM_STYLE = CUSTOM_STYLE;
+app.locals.CUSTOM_STYLE = getCustomStyle();
```
I think the variable CUSTOM_STYLE is no longer necessary.
https://github.com/ResultadosDigitais/matrix/pull/167/files
Or maybe is missing feature in Morpheus layout

alesshh replied:
🔪

juliemar replied:
Yes... we can remove this variable :) my fault :disappointed:
